docs/adr/0005-controller-structure.md

```diff
@@ -0,0 +1,124 @@
+# 5. Controller Structure
+
+Date: 2020-08-18
+
+## Status
+
+Accepted
+
+## Context
+
+We cannot design endpoints in a flat structure due to existing requirements around the Find service. By flat structure, we're referring to just having endpoints like:
+
+- `/providers`
+- `/courses`
+
+Instead, we have to design endpoints in a nested structure scoped to their parent resources, ie:
+
+- `/recruitment_cycles/:recruitment_cycle_id/providers/:provider_code`
+- `/recruitment_cycles/:recruitment_cycle_id/providers/:provider_code/courses/:course_code`.
+
+If we rely on having single controllers to deal with resources which may be nested under other resources, we'll end up mixing concerns/responsibilities of different contraints into one class which will make it difficult to maintain over the long term and less flexible to change. It also violates the single responsibility principle as we end up bloating the class.
+
+## Options
+
+### 1. Single controllers (Default)
+
+The first option is to do nothing and continue to use single controllers.
+
+As an example, given the following routes:
+
+```ruby
+    namespace :public do
+      namespace :v1 do
+        resources :recruitment_cycles, param: :year, only: [:show] do
+          resources :courses, only: %i[index]
+
+          resources :providers, only: %i[index show], param: :code do
+            resources :courses, only: %i[index show]
+          end
+        end
+      end
+    end
+  end
+```
+
+The controllers would match the following structure:
+
+```
+controllers/
+|- api/public/v1/
+|  |- recruitment_cycles_controller.rb
+|  |- courses_controller.rb
+|  |- providers_controller.rb
+|  +- ...
+```
+
+#### Pros
+
+- Simple to grok at first glance without delving into the class
+
+#### Cons
+
+- Could end up with bloated classes which are difficult to maintain and change
+- Mix concerns and responsbilities based on the routing contraint
+
+### 2. Use modules to organise controllers
+
+An alternative is to use modules to organise controller classes based on their concerns.
+
+As an example, given the following routes:
+
+```ruby
+  namespace :public do
+    namespace :v1 do
+      resources :recruitment_cycles, param: :year, only: [:show] do
+        resources :courses, only: %i[index]
+        resources :providers, only: %i[index show], param: :code do
+          scope module: :providers do
```

Comment:
I had to some work to understand the (small) difference between these routes and the one above. Would it be worth adding a comment above line 78 here to explain that `scope` is necessary to get the directory structure right as below?

Reply:
done https://github.com/DFE-Digital/teacher-training-api/pull/1535/commits/930f98ced2c1ede0c1365b3133974c859b39bdfe